src/Security/RolesBuilder/AdminRolesBuilder.php

```diff
@@ -76,27 +78,63 @@ public function addExcludeAdmin(string $exclude): void
 
     public function getRoles(?string $domain = null): array
     {
+        $adminServiceCodes = array_diff($this->pool->getAdminServiceCodes(), $this->excludeAdmins);
+
+        // get groups and admins sort by config
         $adminRoles = [];
-        foreach ($this->pool->getAdminServiceCodes() as $code) {
-            if (\in_array($code, $this->excludeAdmins, true)) {
-                continue;
-            }
+        foreach ($this->pool->getAdminGroups() as $groupCode => $group) {
+            foreach ($group['items'] as $item) {
+                if (!isset($item['admin'])) {
+                    continue;
+                }
+
+                $key = array_search($item['admin'], $adminServiceCodes, true);
+                if (false === $key) {
+                    continue;
+                }
+                unset($adminServiceCodes[$key]);
 
-            $admin = $this->pool->getInstance($code);
-            $securityHandler = $admin->getSecurityHandler();
-            $baseRole = $securityHandler->getBaseRole($admin);
-            foreach (array_keys($admin->getSecurityInformation()) as $key) {
-                $role = sprintf($baseRole, $key);
-                $adminRoles[$role] = [
-                    'role' => $role,
-                    'label' => $key,
-                    'role_translated' => $this->translateRole($role, $domain),
-                    'is_granted' => $this->isMaster($admin) || $this->authorizationChecker->isGranted($role),
-                    'admin_label' => $admin->getTranslator()->trans($admin->getLabel() ?? ''),
-                ];
+                $groupLabelTranslated = $this->translator->trans($group['label'], [], $group['translation_domain']);
+
+                $adminRoles = array_merge($adminRoles, $this->getAdminRolesByAdminCode($item['admin'], $domain, $groupLabelTranslated, $groupCode));
```

Comment:
First test fail in this line.
TypeError: Argument 3 passed to must be of the type string, null given.